### PR TITLE
[MO] - [clients] -> recursive builder for all clients

### PR DESCRIPTION
**Pull request by see-quick**
### Type of change

- Enhancement / new feature
- Refactoring

### Description

This PR adding a basic client to be able to deploy it in a fluent way. Refactoring the whole suite in that way.

TODO:
```
1) Generic builder for all external clients -> done
2) Internal clients properties builder -> done 
3) External clients properties builder -> done
4) Rebasing 100x times -> done
```

###2nd PR will be concentrate on
- create default client properties for send and recv when `.withKafkaClientProperties()` is not specify 
- instead of passing new KafkaClientProperties() in each client use instance member 

### Checklist

- [x] Check InternalKafkaClientBuilder is working
- [x] Check BasicExternalKafkaClientBuilder is working
- [x] Check OauthExternalKafkaClientBuilder is working
- [x] Check TracingExternalKafkaClientBuilder is working
- [x] Make sure all tests pass

### CODE REVIEW

**Comment by Seequick1** [comment]
@strimzi-ci run test_only

**Comment by Seequick1** [comment]
@strimzi-ci run tests test_only

**Comment by Seequick1** [comment]
@strimzi-ci run tests test_only

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=regression

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=regression

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=regression testcase=KafkaST test_only

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=regression testcase=KafkaST test_only

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=regression testcase=KafkaST test_only

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=regression testcase=KafkaST test_only

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=internalclients,externalclients test_only

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=internalclients,externalclients test_only

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/ListenersST.java [hunk 1]
```diff
@@ -540,40 +741,48 @@ void testCustomCertLoadBalancerAndTlsRollingUpdate() throws Exception {
...
-                    .withNewBrokerCertChainAndKey()
-                        .withSecretName(customCertServer1)
-                        .withKey("ca.key")
-                        .withCertificate("ca.crt")
-                    .endBrokerCertChainAndKey()
+                .withNewBrokerCertChainAndKey()
+                .withSecretName(customCertServer1)
+                .withKey("ca.key")
+                .withCertificate("ca.crt")
+                .endBrokerCertChainAndKey()
```
Comment (Frawless) [inline systemtest/src/test/java/io/strimzi/systemtest/kafka/ListenersST.java]: WHy did you change the indent?

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/KafkaClientProperties.java [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+        }
+
+        public KafkaClientPropertiesBuilder withAcksConfig(String acksConfig) {
+
+            this.properties.setProperty(ProducerConfig.ACKS_CONFIG, acksConfig);
+            return this;
+        }
+
+        // TODO: do we need separate Producer and Consumer config ??
+        public KafkaClientPropertiesBuilder withGroupIdConfig(String groupIdConfig) {
```
[the remark this answers is not in the document]
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/KafkaClientProperties.java]: That's why i write commend here `// TODO: do we need separate Producer and Consumer config ??` and i have three alternatives for this....
```
1) we can create two inner classes  inside this whole class so we will have following hierarchy
 - KafkaClientProperties
      - KafkaClientConsumerProperties
      - KafkaClientProducerProperties
2) create two separate classes 
3) create abstract class and two which will extend from that....
```
Anyway i would change it after this huge PR.

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/ListenersST.java (hunk 1 above)
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/kafka/ListenersST.java]: sorry...copy paste make sometimes problems )))

File: systemtest/src/test/java/io/strimzi/systemtest/ConnectS2IST.java [hunk 3]
```diff
@@ -209,9 +211,17 @@ void testSecretsWithKafkaConnectS2IWithTlsAndScramShaAuthentication() {
...
+            ResourceManager.kubeClient().listPodsByPrefixInName(CLUSTER_NAME + "-" + Constants.KAFKA_CLIENTS).get(0).getMetadata().getName();
+
+        InternalKafkaClient internalKafkaClient = new InternalKafkaClient.Builder()
+            .withUsingPodName(defaultKafkaClientsPodName)
+            .withTopicName(CONNECT_S2I_TOPIC_NAME)
+            .withNamespaceName(NAMESPACE)
+            .withClusterName(CLUSTER_NAME)
+            .withMessageCount(MESSAGE_COUNT)
+            .withSecurityProtocol("TLS")
+            .withConsumerGroupName("my-group-" + new Random().nextInt(Integer.MAX_VALUE))
```
[the remark this answers is not in the document]
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/ConnectS2IST.java]: sure, we can create some issue for that as enhancement ))

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=internalclients,externalclients test_only

File: systemtest/src/test/java/io/strimzi/systemtest/ConnectS2IST.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/ConnectS2IST.java]: anyway, it would be better just to encapsulate this `consumer-group-name` and have two options:
1) if specified create that with the concrete name
2) if not specify create the random one

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=internalclients,externalclients test_only

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=internalclients,externalclients test_only

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=internalclients,externalclients test_only

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=internalclients,externalclients test_only

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/KafkaClientProperties.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Frawless) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/KafkaClientProperties.java:114]: Make sense to leave it for another PR. We can offline discuss the best solution

File: systemtest/src/test/java/io/strimzi/systemtest/ConnectS2IST.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/ConnectS2IST.java:230]: Agree

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/internalClients/VerifiableClient.java [hunk 4]
```diff
@@ -84,6 +182,18 @@ public void setArguments(ClientArgumentMap args) {
         }
     }
 
+    private boolean allowParameter(String minimalVersion, String clientVersion) {
```
Comment (Frawless) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/internalClients/VerifiableClient.java:185]: Do we need this now?

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/internalClients/VerifiableClient.java [hunk 5]
```diff
@@ -28,21 +30,117 @@
...
+    private ClientType clientType;
     private String podName;
     private String podNamespace;
+    private String bootstrapServer;
+    private String topicName;
+    private int maxMessages;
+    private String kafkaUsername;
+    private String consumerGroupName;
+    private String consumerInstanceId;
+    // TODO: verbose ?? consumerArguments.put(ClientArgument.VERBOSE, "");
```
Comment (Frawless) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/internalClients/VerifiableClient.java:44]: Remove or?

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/internalClients/VerifiableClient.java (hunk 4 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/internalClients/VerifiableClient.java:185]: spot-bugs already did this spot :D but nice catch 👍

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=internalclients,externalclients testcase=io.strimzi.systemtest.oauth.OauthPlainST#testProducerConsumerMirrorMaker,io.strimzi.systemtest.oauth.OauthPlainST#testProducerConsumerBridge{Vertx},io.strimzi.systemtest.oauth.OauthPlainST#testIntrospectionEndpointWithPlainCommunication,io.strimzi.systemtest.oauth.OauthPlainST#testProducerConsumer,io.strimzi.systemtest.oauth.OauthPlainST#testProducerConsumerMirrorMaker2,io.strimzi.systemtest.oauth.OauthPlainST#testProducerConsumerConnect,io.strimzi.systemtest.kafka.ListenersST#testCustomCertLoadBalancerAndTlsRollingUpdate,io.strimzi.systemtest.kafka.ListenersST#testCustomSoloCertificatesForLoadBalancer,io.strimzi.systemtest.MirrorMaker2ST#testMirrorMaker2TlsAndScramSha512Auth,io.strimzi.systemtest.RollingUpdateST#testRecoveryDuringZookeeperRollingUpdate,io.strimzi.systemtest.MirrorMakerST#testMirrorMakerTlsScramSha,io.strimzi.systemtest.ConnectST#testSecretsWithKafkaConnectWithTlsAndScramShaAuthentication,io.strimzi.systemtest.SecurityST#testCaRenewalBreakInMiddle

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        // for not explicit casting..
+        protected abstract T self();
+    }
+
+    protected AbstractKafkaClient(Builder<?> builder) {
+
+        if (builder.topicName == null) throw new InvalidParameterException("Topic name is not set.");
+        if (builder.namespaceName == null) throw new InvalidParameterException("Namespace name is not set.");
+        if (builder.clusterName == null) throw  new InvalidParameterException("Cluster name is not set.");
+        if (builder.messageCount == 0) throw  new InvalidParameterException("Message count is set to 0");
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java:98]: What if it's negative?

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        // Subclasses must override this method to return "this" protected abstract T self();
+        // for not explicit casting..
+        protected abstract T self();
+    }
+
+    protected AbstractKafkaClient(Builder<?> builder) {
+
+        if (builder.topicName == null) throw new InvalidParameterException("Topic name is not set.");
+        if (builder.namespaceName == null) throw new InvalidParameterException("Namespace name is not set.");
+        if (builder.clusterName == null) throw  new InvalidParameterException("Cluster name is not set.");
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java:97]: What if they're empty?

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+    protected String topicName;
+    protected String namespaceName;
+    protected String clusterName;
+    protected int messageCount;
+    protected String consumerGroup;
+    protected String kafkaUsername;
+    protected String securityProtocol;
+    protected String caCertName;
+    protected KafkaClientProperties clientProperties;
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java:27]: You've got a builder, so why can't these be `final`?

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    }
+    /**
+     * Get external bootstrap connection
+     * @param namespace kafka namespace
+     * @param clusterName kafka cluster name
+     * @return bootstrap url as string
+     */
+    protected static String getExternalBootstrapConnect(String namespace, String clusterName) {
+        if (kubeClient(namespace).getClient().isAdaptable(OpenShiftClient.class)) {
+            Route route = kubeClient(namespace).getClient().adapt(OpenShiftClient.class).routes().inNamespace(namespace).withName(clusterName + "-kafka-bootstrap").get();
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java:126]: So how would you test non-`Route` external connections on OpenShift?

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+        } else if (extBootstrapServiceType.equals("LoadBalancer")) {
+            LoadBalancerIngress loadBalancerIngress = extBootstrapService.getStatus().getLoadBalancer().getIngress().get(0);
+            String result = loadBalancerIngress.getHostname();
+
+            if (result == null) {
+                result = loadBalancerIngress.getIp();
+            }
+            return result + ":9094";
+        } else {
+            throw new RuntimeException("Unexpected external bootstrap service for Kafka cluster " + clusterName);
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java:156]: Don't you want to mention what the unexpected `extBootstrapServiceType` was?

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/IKafkaClientOperations.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.systemtest.kafkaclients;
+
+public interface IKafkaClientOperations<T> {
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/IKafkaClientOperations.java:7]: Why are we naming interfaces with `I` prefix here, when it's not something we do in the rest of the code base? Can't you think of a better name without the `I`?

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/KafkaClientProperties.java [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+
+/**
+ * Class KafkaClientProperties, which holds inner class builder for fluent way to invoke objects. It is used inside
+ * all our external clients such as BasicExternalKafkaClient or OauthExternalKafkaClient.
+ *
+ * Description of SuppressWarnings:
+ *
+ * caSecretName field is not initialized first inside KafkaClientProperties constructor and he is de-referenced
+ * in sharedClientProperties() method. This practically means, always make sure that before invoking this method
+ * you need first execute withCaSecretName().
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/KafkaClientProperties.java:56]: * There's no need to document these it in a javadoc comment, a normal comment will suffice. 
* The reasons can get stale. The interested reader can presumable try removing the suppressions and see what warnings they get and why.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/KafkaClientProperties.java [hunk 13]
```diff
@@ -0,0 +1,458 @@
...
+            try {
+                if (!properties.getProperty(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG).equals(CommonClientConfigs.DEFAULT_SECURITY_PROTOCOL) &&
+                    !properties.getProperty(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG).equals("SASL_" + CommonClientConfigs.DEFAULT_SECURITY_PROTOCOL)
+                ) {
+                    Secret clusterCaCertSecret = kubeClient(namespaceName).getSecret(caSecretName);
+                    File tsFile = File.createTempFile(KafkaClientProperties.class.getName(), ".truststore");
+                    String tsPassword = "foo";
+                    KeyStore ts = KeyStore.getInstance(TRUSTSTORE_TYPE_CONFIG);
+                    if (caSecretName.contains("custom-certificate")) {
+                        tsFile.deleteOnExit();
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/KafkaClientProperties.java:247]: since we should always delete these it's better to remove from both branches and put it immediately after it's creation.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/BasicExternalKafkaClient.java [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+        CompletableFuture<Integer> resultPromise = new CompletableFuture<>();
+
+        IntPredicate msgCntPredicate = x -> x == messageCount;
+
+        vertx.deployVerticle(new Producer(
+            new KafkaClientProperties.KafkaClientPropertiesBuilder()
+                .withNamespaceName(namespaceName)
+                .withClusterName(clusterName)
+                .withBootstrapServerConfig(getExternalBootstrapConnect(namespaceName, clusterName))
+                .withKeySerializerConfig(StringSerializer.class.getName())
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/BasicExternalKafkaClient.java:70]: Why does the `withKeySerializerConfig()` class take a `String` if it could take a `Class`?

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/IKafkaClientOperations.java [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+
+    T sendMessagesPlain(long timeoutMs) throws Exception;
+    T sendMessagesTls(long timeoutMs) throws Exception;
+
+    T receiveMessagesPlain(long timeoutMs) throws Exception;
+    T receiveMessagesTls(long timeoutMs) throws Exception;
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/IKafkaClientOperations.java:13]: If you're going to the trouble of abstracting an interface it's probably worth documenting the semantics of these methods.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+        String caCertName = this.caCertName == null ?
+                KafkaResource.getKafkaExternalListenerCaCertName(namespaceName, clusterName) : this.caCertName;
+        LOGGER.info("Going to use the following CA certificate: {}", caCertName);
+
+        KafkaClientProperties kafkaClientProperties = new KafkaClientProperties.KafkaClientPropertiesBuilder()
+            .withNamespaceName(namespaceName)
+            .withClusterName(clusterName)
+            .withCaSecretName(caCertName)
+            .withBootstrapServerConfig(getExternalBootstrapConnect(namespaceName, clusterName))
+            .withKeyDeSerializerConfig(StringDeserializer.class.getName())
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java:233]: `withKeyDeSerializerConfig` → `withKeyDeserializerConfig`

Also if everything is just going to use `StringDeserializer` why not make that the default so you only need to call the method if you're using a different Deserializer?

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+            .withValueDeSerializerConfig(StringDeserializer.class.getName())
+            .withKafkaUsername(kafkaUsername)
+            .withSecurityProtocol("SASL_SSL")
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java:236]: You could make the method take `org.apache.kafka.common.security.auth.SecurityProtocol` for better type safety, if you don't mind the dependency leaking to your clients.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+            .withGroupIdConfig(consumerGroup)
+            .withAutoOffsetResetConfig("earliest")
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java:238]: Similarly you could use `org.apache.kafka.clients.consumer.OffsetResetStrategy`

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/TracingExternalKafkaClient.java [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Override
+    public Future<Integer> receiveMessagesPlain(long timeoutMs) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Future<Integer> receiveMessagesTls(long timeoutMs) {
+        throw new UnsupportedOperationException();
+    }
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/TracingExternalKafkaClient.java:69]: Something is wrong if you're implementing an interface with all methods throwing `UnsupportedOperationException`. I'm _guessing_ you've had to do this because you wanted a different builder for the tracing case? I think you need to question why your builder has to be an inner class of some `AbstractKafkaClient` subclass.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java (hunk 6 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java:98]: fair point )

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java (hunk 8 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java:27]: yeah, that's a problem that come in my mind. If you take a close look on how I am using the client for instance:
```
InternalKafkaClient internalKafkaClient = new InternalKafkaClient.Builder()
            .withUsingPodName(kafkaClientsPodName)
            .withTopicName("my-topic-test-3")
            .withNamespaceName(NAMESPACE)
            .withClusterName(kafkaClusterSourceName)
            .withKafkaUsername(userSource.getMetadata().getName())
            .withMessageCount(messagesCount)
            .withConsumerGroupName(CONSUMER_GROUP_NAME + "-" + rng.nextInt(Integer.MAX_VALUE))
            .build();
```

and then after a few lines, I need to set some of the attributes to the new value

```
internalKafkaClient.setClusterName(kafkaClusterTargetName); internalKafkaClient.setKafkaUsername(userTarget.getMetadata().getName());
```
and that's the problem why these properties some of them can't be final. Sure I can rewrite some of them to the final, which is not changed by the setter but that I consider as a not proper way :)

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/BasicExternalKafkaClient.java (hunk 14 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/BasicExternalKafkaClient.java:70]: Yeah, I can do it to use Class instead of passing Strings thanks!

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java (hunk 16 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java:233]:
```
Also if everything is just going to use StringDeserializer why not make that the default so you only need to call the method if you're using a different Deserializer?
```
I can create the next PR. on that.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java (hunk 17 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java:236]: Great idea!! I was wondering about the Enum, but this one solves it immediately.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java (hunk 18 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java:238]: This will be questioned for Serialization and also this one. What gives me that I can provide `OffsetResetStrategy.class`? It is just of type compatibility and the wreck of more exuberance of semanticism?  I totally agree it is better to use classes instead of literals like this, but from the performance POV it is easy to just provide string right not just whole class?

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java (hunk 9 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java:126]: `non-Route` i am not quite sure that I fully understand this statement.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/TracingExternalKafkaClient.java (hunk 19 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/TracingExternalKafkaClient.java:69]: Practically, this should be implemented in a few weeks. It will be the Tracing clients using these methods and yeah.

**Comment by see-quick** [comment]
@tombentley  so `AbstractKafkaClient` and inner builder is for them as you are guessing correctly for differences between the Oauth and tracing clients, which for instance OAuth uses endpoint token URI and so on. Moreover,  tracing clients using service-name for the Jaeger so it would be not smart to put these properties inside one BIG builder, which will hold them all but instead of this I create these three separate sub-classes classes, which shares common properties name-space, cluster-name and so on.....

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=internalclients,externalclients test_only

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java (hunk 8 above)
Comment (tombentley) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java]: It seems strange to me that you want the client first to point to one cluster and then another. It would be more natural if a client, once created, was immutable and you use two clients instance. What you need is to be able to base a builder's state on a pre-existing client, and then tweak it with a few more `with()` calls, so you can build the 2nd client conveniently.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java (hunk 9 above)
Comment (tombentley) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java:128]: Well that code basically means you can _only_ test Routes on OpenShift. You can't test NodePorts, for instance.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java (hunk 18 above)
Comment (tombentley) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java:238]: I think you misunderstood my comment. The argument provided by the caller in this example would be `OffsetResetStrategry.EARLIEST`. For the serializer you'd need to provide a `Class<? extends Serializer>`. Different cases of the same general point, which is that it's not very type safe to use Strings for these things, and there are existing types you can easily use to provide more type safety to your API.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/TracingExternalKafkaClient.java (hunk 19 above)
Comment (tombentley) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/TracingExternalKafkaClient.java:69]: OK. A `// TODO` or exception message to that effect is always helpful to show the intent to come back and address it.

**Comment by tombentley** [comment]
> Moreover, tracing clients using service-name for the Jaeger so it would be not smart to put these properties inside one BIG builder, which will hold them all but instead of this I create these three separate sub-classes classes, which shares common properties name-space, cluster-name and so on.....

OK, that explains your reasoning, but I'm not sure I agree that this should be reflected in the API like this. That's because it makes it difficult to build a client which does both tracing and oauth. That might not look like a compelling argument today, because maybe we can't imagine needing to test that configuration. However, it's likely that in the future we'll end up with more of these "adaptors", which give you a non-vanilla client, and the more such cases we have the more harmful it is that we can't have a client which is both an X and a Y client. The fact is that these are not disjoint things, so it's a mistake to create an API which forces them to be disjoint. 

As for a solution... would nested builders work?

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java (hunk 18 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java:238]: Yeah, got it. Maybe last question using of`Class<? extends Serializer>` would be using this class in the raw. Would it be still ok instead of specifying concrete type Serializer<T>. In that case `T will be StringSerializer` i assume

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java (hunk 9 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java:128]: Hmmm, you can test Nodeports the same with the LoadBalancers. The tests will all failed if this method is not working. 
see
https://github.com/strimzi/strimzi-kafka-operator/pull/2520/files?file-filters%5B%5D=.java#diff-b75f4b3368b48a1b5ad2ccd692700010R145-R156

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java (hunk 8 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java]: In the next PR, I want to create some `default clients`, which will satisfy it. Also, I agree, with immutalibity and will do it.

**Comment by see-quick** [comment]
Nested builders will work in my opinion. I agree, with your arguments that someday if we need to test the combination of these clients with the current configuration it will not work. However, we can look at this problem with the positive side and it is when if in the future will come the day when we need to test OAuth + Tracing + X + Y + Z, we can implement our client based on the AbstractKlient. For instance, we can implement OAuth + tracing client which will basically add some redundancy but w/e it will work.

**Comment by see-quick** [comment]
@tombentley anyway I am sure that this is not the best solution but the currently better solution is not incoming in my brain :D

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java (hunk 8 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java]: I have already created a task for that.

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=internalclients,externalclients testcase=io.strimzi.systemtest.kafka.ListenersST#testCustomCertLoadBalancerAndTlsRollingUpdate,io.strimzi.systemtest.MirrorMaker2ST#testMirrorMaker2TlsAndScramSha512Auth,io.strimzi.systemtest.SecurityST#testCaRenewalBreakInMiddle

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=acceptance tests_only

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=acceptance tests_only

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=acceptance tests_only

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=acceptance tests_only

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=acceptance tests_only

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=acceptance testcase=io.strimzi.systemtest.ConnectST#testKafkaConnectAndConnectorFileSinkPlugin tests_only

**Comment by see-quick** [comment]
So, Travis if you check in detail. If there is no review from @ppatierno i would merge it as soon as possible. @scholzj . Many tasks is dependent on this one )))
